File: simple-stack-example-fragments/src/main/java/com/zhuinden/simplestackdemoexamplefragments/application/CustomApplication.java
package com.zhuinden.simplestackdemoexamplefragments.application;

import android.app.Application;
import android.content.Context;

import com.zhuinden.simplestackdemoexamplefragments.application.injection.DaggerSingletonComponent;
import com.zhuinden.simplestackdemoexamplefragments.application.injection.SingletonComponent;

/**
 * Created by Owner on 2017. 01. 26..
 */

public class CustomApplication
        extends Application {
    SingletonComponent singletonComponent;

    private static CustomApplication INSTANCE;

    public void initialize() {
        if(INSTANCE == null) {
            INSTANCE = this;
            singletonComponent = DaggerSingletonComponent.create();
        }
    }

    public static CustomApplication get(Context context) {
        return (CustomApplication) context.getApplicationContext();
    }

    public SingletonComponent getComponent() {
        return singletonComponent;
    }

    public static CustomApplication get() {
        return INSTANCE;
    }
}

File: simple-stack-example-mvp-fragments/src/main/java/com/zhuinden/simplestackdemoexamplefragments/application/CustomApplication.kt
package com.zhuinden.simplestackdemoexamplefragments.application

import android.app.Application
import android.content.Context

import com.zhuinden.simplestackdemoexamplefragments.application.injection.DaggerSingletonComponent
import com.zhuinden.simplestackdemoexamplefragments.application.injection.SingletonComponent

/**
 * Created by Zhuinden on 2018. 08. 20.
 */

class CustomApplication : Application() {
    companion object {
        private var INSTANCE: CustomApplication? = null

        private var component: SingletonComponent? = null

        @JvmStatic
        fun getComponent() = component!!

        @JvmStatic
        fun get(context: Context): CustomApplication {
            return context.applicationContext as CustomApplication
        }

        @JvmStatic
        fun get(): CustomApplication = INSTANCE!!
    }

    override fun onCreate() {
        super.onCreate()
        INSTANCE = this
        component = DaggerSingletonComponent.create()
    }
}
